### PR TITLE
[FSDP] Provide a utility API to allow users easily to set state_dict_type

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73716

fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -760,12 +760,15 @@ def _need_prefetch_post_backward_hook(self) -> bool:
         else:
             return False
 
+    @staticmethod
     @contextlib.contextmanager
-    def state_dict_type(self, state_dict_type: StateDictType) -> Generator:
+    def state_dict_type(module: nn.Module, state_dict_type: StateDictType) -> Generator:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:765]: maybe rename it to `fsdp_state_dict_type`, unless you feel that the call `FSDP.state_dict_type()` is already clear enough that it's setting things for only FSDP modules.

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 2]
```diff
@@ -197,7 +197,7 @@ def _load_state_dict(
         except KeyError:
             raise ValueError(f"No state_dict for {state_dict_type}")
 
-        with model.state_dict_type(enum_val):
+        with model.state_dict_type(model, enum_val):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_state_dict.py:200]: why not `FSDP.state_dict_type`?

**Review by rohan-varma — APPROVED** [review]
LGTM, just some minor comments and questions

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:765]: I feel `FSDP.state_dict_type()` explicit enough. Only FSDP internal code will use something like `module.state_dict_type()` or `self.state_dict_type()`.

File: test/distributed/fsdp/test_fsdp_state_dict.py (hunk 2 above)
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_state_dict.py:200]: Good catch!

**Pull request closed without merging**